text/0000-pattern-3.md

```diff
@@ -0,0 +1,1653 @@
+where
+    H: Haystack,
+    P: Pattern<H>;
+
+pub fn rfind_range<H, P>(haystack: H, pattern: P) -> Option<Range<H::Target::Index>>
+where
+    H: Haystack,
+    P: Pattern<H, Searcher: ReverseSearcher<H::Target>>;
+```
+
+**Split**
+
+```rust
+pub fn split<H, P>(haystack: H, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H>;
+
+pub fn rsplit<H, P>(haystack: H, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H, Searcher: ReverseSearcher<H::Target>>;
+
+pub fn split_terminator<H, P>(haystack: H, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H>;
+
+pub fn rsplit_terminator<H, P>(haystack: H, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H, Searcher: ReverseSearcher<H::Target>>;
+
+pub fn splitn<H, P>(haystack: H, n: usize, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H>;
+
+pub fn rsplitn<H, P>(haystack: H, n: usize, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H, Searcher: ReverseSearcher<H::Target>>;
+```
+
+**Replace**
+
+```rust
+pub fn replace_with<H, P, F, W>(src: H, from: P, replacer: F, writer: W)
+where
+    H: Haystack,
+    P: Pattern<H>,
+    F: FnMut(H) -> H,
+    W: FnMut(H);
+
+pub fn replacen_with<H, P, F, W>(src: H, from: P, replacer: F, n: usize, writer: W)
+where
+    H: Haystack,
+    P: Pattern<H>,
+    F: FnMut(H) -> H,
+    W: FnMut(H);
+```
+
+</details>
+
+Most algorithms are very simple to implement using trisection (`.split_around()`). For instance,
+`split()` can be implemented as:
+
+```rust
+gen fn split<H, P>(haystack: H, pattern: P) -> impl Iterator<Item = H>
+where
+    H: Haystack,
+    P: Pattern<H>,
+{
+    let mut searcher = pattern.into_searcher();
+    let mut rest = Span::from(haystack);
+    while let Some(range) = searcher.search(rest.borrow()) {
+        let [left, _, right] = unsafe { rest.split_around(range) };
+        yield left.into();
+        rest = right;
+    }
+    yield rest;
+}
+```
+
+These functions are forwarded as *inherent methods* of the haystack type, e.g.
+
+```rust
+impl str {
+    ...
+
+    pub fn split_mut<'a>(
+        &'a mut self,
+        pattern: impl Pattern<&'a mut str>,
+    ) -> impl Iterator<Item = &'a mut str> {
+        core::pattern::split(self, pattern)
+    }
+
+    pub fn replace<'a>(
+        &'a self,
+        from: impl Pattern<&'a str>,
+        to: &str,
+    ) -> String {
+        let mut res = String::with_capacity(self.len());
+        core::pattern::replace_with(self, from, |_| to, |r| res.push_str(r));
+        res
+    }
+
+    ...
+}
+```
+
+## Standard library changes
+
+* Remove the entire `core::str::pattern` module from public, as this is unstable.
+
+* Add the `core::pattern` module with traits and structs shown above.
+
+* Implement `Hay` to `str`, `[T]` and `OsStr`.
+
+* Implement `Haystack` to `∀H: Hay. &H`, `&mut str` and `&mut [T]`.
```

Comment:
The `pattern_3` crate also has an implementation for `Vec<T>` (though not `String` or `OsString`). Are those owned implementations intended eventually? Is that just out of scope for this particular RFC?

Reply:
I don't intend to add these into the standard library, due to the efficiency concern you've raised. 

`pattern_3` does implement for `Vec<T>` just to illustrate that it can transfer owned data type correctly.